In [41]:
%%capture
# Load SQL extension and connect to DuckDB
%load_ext sql
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False
%sql duckdb://

### **Monthly Revenue Performance**

How has monthly revenue changed over time, and how many orders and customers are driving that revenue?

In [51]:
%%sql

WITH monthly_revenue AS (
SELECT 
    strftime(order_date, '%Y-%m') AS month,
    COUNT(order_id) AS total_orders,
    COUNT(DISTINCT customer_id) AS unique_customers,
    ROUND(SUM(revenue), 0)::INT AS total_revenue
FROM 'data/orders.csv'
WHERE order_status = 'Completed'
GROUP BY month
ORDER BY month
)

SELECT
    *
FROM monthly_revenue
WHERE month = (SELECT MIN(month) FROM monthly_revenue) OR month = (SELECT MAX(month) FROM monthly_revenue)
;

month,total_orders,unique_customers,total_revenue
2024-01,42,41,32088
2026-08,1939,1166,1466817


In the past 2.5 years, Northstar has grown from 41 customers and $32,088 in revenue to 1,166 customers and $1,466,817 in revenue. 


********


### Revenue by Acquisition Channel

Which customer acquisiton channels generate the most in total revenue?

In [54]:
%%sql

SELECT 
    RANK() OVER (ORDER BY SUM(o.revenue) DESC) AS rank,
    mc.channel_name AS channel_name,
    ROUND(SUM(o.revenue), 0)::INT AS total_revenue
FROM 'data/orders.csv' o
LEFT JOIN 'data/customers.csv' c
    ON o.customer_id = c.customer_id
LEFT JOIN 'data/marketing_channels.csv' mc
    ON c.channel_id = mc.channel_id
WHERE o.order_status = 'Completed'
GROUP BY channel_name
ORDER BY total_revenue DESC;

rank,channel_name,total_revenue
1,Organic,6764476
2,Google Ads,4503094
3,Referral,4450311
4,Outbound Sales,3622330
5,Partner,1313463
6,Social,957119
7,Content,647671
8,Events,458620


Organic channels generate the most in revenue whereas Events generates the least.


--------


### Top 25 Customers by lifetime revenue


In [ ]:
%%sql

SELECT *
FROM 'data/customers.csv' c


SyntaxError: invalid syntax (1630049499.py, line 1)